### ISSUE TITLE
Add warning and error options, like eslint-webpack-plugin and eslint-loader

### ISSUE THREAD

**Issue by DamianPereira** [TypeStrong/fork-ts-checker-webpack-plugin#514]
## Feature motivation

Both [eslint-webpack-plugin](https://github.com/webpack-contrib/eslint-webpack-plugin) and [eslint-loader](https://github.com/webpack-contrib/eslint-loader) provide options to control warnings and error emitting. For example emitError will emit warnings as errors, and failOnWarning will fail the build if there are warnings.

I'd like to have webpack dev server only fail on errors on development (as it works by default). But I'd also like to have the production build fail if there is any warnings. 

## Feature description

[These options](https://github.com/webpack-contrib/eslint-loader) would have to be added to the eslint object. Note that these are *not* eslint compiler options, like the ones we can provide in the options object, but rather options for the plugin to treat eslint errors and warnings differently.

## Feature implementation

I'm not sure where exactly this should be added, but conditions like [these](https://github.com/TypeStrong/fork-ts-checker-webpack-plugin/search?q=compilation.warnings.push&unscoped_q=compilation.warnings.push) will probably need to be changed so that, for example, if emitError is true, warnings result in compiler.errors.push(warning), and similar code for the other combinations.

**Comment by piotr-oles** [comment]
There is an `issues` plugin hook that you can use for a custom use-case.

```typescript
import webpack from 'webpack';
import ForkTsCheckerWebpackPlugin from 'fork-ts-checker-webpack-plugin';

interface MyPluginThatModifiesEmittedIssuesOptions {}

class MyPluginThatModifiesEmittedIssues implements webpack.Plugin {

  constructor(private options: MyPluginThatModifiesEmittedIssuesOptions) {}

  apply(compiler: webpack.Compiler) {
     const hooks = ForkTsCheckerWebpackPlugin.getCompilerHooks(compiler);
     hooks.issues.tap((issues) => {
       // do whatever you want with your issues
       return issues;
     });
  }
}
```
Usage:
```javascript
// webpack.config.js

module.exports = {
  // ...
  plugins: [
    new ForkTsCheckerWebpackPlugin(),
    new MyPluginThatModifiesEmittedIssues({
      // options you want to pass to the plugin
    })
  ]
}
```
I wrote this code from memory, so not sure if everything is ok, but it should point you in the right direction :)

**Comment by DamianPereira** [comment]
Is there any documentation on what is inside that issues parameter? Can I stop the files compiled by typescript and checked by eslint from being emitted by interacting with that issues object?

**Comment by DamianPereira** [comment]
I managed to replace the emitError option with your example, doing this
```javascript
class TreatTSWarningsAsErrorPlugin {
  constructor(enable) {
    this.enable = enable;
  }

  apply(compiler) {
    const hooks = ForkTsCheckerWebpackPlugin.getCompilerHooks(compiler);
    hooks.issues.tap('TreatTSWarningsAsError', (issues) => {
      return this.enable
        ? issues.map((issue) => ({ ...issue, severity: 'error', message: `WARNING: ${issue.message}` }))
        : issues;
    });
  }
}
```

But I can't replace failOnWarnings. Whenever there are errors, warnings are hidden, so if I throw inside apply, or add an error to the issues array, it won't show the warnings, so you don't know what to correct. I'd need a way to show warnings even if there are errors.

**Comment by piotr-oles** [comment]
Hmmm... There is no special logic inside the plugin that ignores warnings if there are errors. Do you use async: false mode? If yes, I would bet that this logic is inside webpack. So maybe you would have to tweak webpack's 'stats' configuration.

**Comment by DamianPereira** [comment]
I've been investigating it, and seems like the eating up warnings on errors is caused by https://github.com/geowarin/friendly-errors-webpack-plugin. a failOnWarnings would still be a good option.

The alternative with the compiler hooks would be to add an error when there are any warnings, but it's still not as usable.

**Comment by piotr-oles** [comment]
In async: true, the plugin is not using webpack's infrastructure to print errors - it just console.logs errors. If FriendlyErrorsWebpackPlugin clears the screen, it's possible that it eats up these warnings/errors :)
The failOnWarnings option seems to be outside of the plugin's domain. I don't want to add more and more options because it will increase the maintenance cost of the plugin. It seems like something that should be done on the webpack side. If you disagree and you think it would improve the developer experience a lot, I'm open to a PR, but I will not guarantee that it will be merged if it adds too much complexity :/

**Comment by DamianPereira** [comment]
I will do some more experiments, and see if this is something I could do and provide a PR (and if it adds too much complexity). But I probably won't be able to work on this for a while. Do you mind leaving this open in the meantime?

**Comment by piotr-oles** [comment]
Sure, no problem :)

**Comment by piotr-oles** [comment]
Closing due to inactivity :)
